### ISSUE TITLE
Lossless JPEG: Add support for decoding images with 14-bit data precision (needed by DNG)

### ISSUE THREAD

**Issue by SimonSegerblomRex** [libjpeg-turbo/libjpeg-turbo#768]
According to the JPEG specification all values between 2 and 16 are valid. The first thing to fix in libjpeg-turbo is `jpeg_read_header` which fails on any values other than 8, 12, and 16.

Here's the minimal effort required:
```diff
diff --git a/jdapistd.c b/jdapistd.c
index 1f449272..f04a97a4 100644
--- a/jdapistd.c
+++ b/jdapistd.c
@@ -313,7 +313,7 @@ _jpeg_read_scanlines(j_decompress_ptr cinfo, _JSAMPARRAY scanlines,
 #if BITS_IN_JSAMPLE != 16 || defined(D_LOSSLESS_SUPPORTED)
   JDIMENSION row_ctr;
 
-  if (cinfo->data_precision != BITS_IN_JSAMPLE)
+  if (cinfo->data_precision > BITS_IN_JSAMPLE)
     ERREXIT1(cinfo, JERR_BAD_PRECISION, cinfo->data_precision);
 
   if (cinfo->global_state != DSTATE_SCANNING)
diff --git a/jdinput.c b/jdinput.c
index 136bef59..ea30a382 100644
--- a/jdinput.c
+++ b/jdinput.c
@@ -58,8 +58,7 @@ initial_setup(j_decompress_ptr cinfo)
 
   /* For now, precision must match compiled-in value... */
 #ifdef D_LOSSLESS_SUPPORTED
-  if (cinfo->data_precision != 8 && cinfo->data_precision != 12 &&
-      cinfo->data_precision != 16)
+  if (!(2 <= cinfo->data_precision && cinfo->data_precision <= 16))
 #else
   if (cinfo->data_precision != 8 && cinfo->data_precision != 12)
 #endif
```

I think this would be good enough as a start. Then it becomes possible to decode any Lossless JPEG images with:
```c
    ...
    jpeg_stdio_src(&cinfo, infile);
    jpeg_read_header(&cinfo, TRUE);  // failed before
    int bit_depth = cinfo.data_precision;
    cinfo.data_precision = 16;
    jpeg_start_decompress(&cinfo);
    cinfo.data_precision = bit_depth;
    ...
        jpeg16_read_scanlines(&cinfo, ...);
    ...
```
Getting rid of the workaround changing `cinfo.data_precision` requires a lot more changes (that might not add any value).

I realize you might not have time to look into this at the moment, but it's something for the backlog at least.

(BitsPerSample=14 is one of the most common values for Lossless JPEG encoded images in DNG containers.)

**Comment by dcommander** [comment]
Obviously much more effort would be required than just hacking the library to bypass the error. The error serves a valid purpose, because the underlying library cannot actually support data precisions other than 8, 12, or 16. Additional data precisions were never implemented by Tom Lane (the original libjpeg author) or Ken Murchison (the author of the lossless patch against libjpeg.) Integrating lossless support was a massive effort that required a major financial sponsorship from MathWorks. There is no additional funding for implementing other data precisions, nor is there any real demand for that. (Medical imaging was the primary driver of the lossless JPEG feature, and medical images only need 8, 12, or 16 bits.) There are no plans to support data precisions other than 8, 12, and 16 in libjpeg-turbo, and I do not accept PRs for incomplete features or hacks.

**Comment by dcommander** [comment]
Sorry, I missed your last statement. Please post an example image, and I will reopen with a more limited scope of supporting 14-bit only. But I would still need funding to work on it.

**Comment by dcommander** [comment]
Maybe I'm misunderstanding something. Are you saying that the underlying algorithms actually work if there is a precision mismatch? Please explain.

**Comment by SimonSegerblomRex** [comment]
> Maybe I'm misunderstanding something. Are you saying that the underlying algorithms actually work if there is a precision mismatch? Please explain.

Yes! BitsPerSample is only used as an exponent in the encoding process and it would be impossible to implement decoding for only BitsPerSample=16 without also supporting all the lower bit depths. I've verified that I can decode images correctly with all values of BitsPerSample between 2 and 16 with this patch, I'll upload some example images later today or tomorrow.

**Comment by dcommander** [comment]
No rush. I won't be able to look into it for a couple of weeks. Would it also be the case that, with a sample size of 8 or 12 bits, you can correctly decode any lossless JPEG image with that data precision or less? Would also probably need to extend the encoder so it can generate such images, and improve the documentation to highlight the difference between sample sizes in the JPEG image and in the packed-pixel image. And testing will be necessary, but this at least seems doable with the General Fund now.

**Comment by SimonSegerblomRex** [comment]
> Would it also be the case that, with a sample size of 8 or 12 bits, you can correctly decode any lossless JPEG image with that data precision or less?

I think so, but I've only tried using `jpeg16_read_scanlines` so far.

I uploaded the following files to [Google Drive](https://drive.google.com/drive/folders/1ocDXK1bvWSOOZ66Wwp-0pEVhxobLVf2i?usp=sharing):

Three images from https://raw.pixls.us/ with BitsPerSample=14:
* [Sony - ILCE-7RM5 - 14bit (4:3).ARW](https://raw.pixls.us/getfile.php/6237/nice/Sony%20-%20ILCE-7RM5%20-%2014bit%20(4:3).ARW)
* [Canon - EOS-1D X - RAW (3:2).CR2](https://raw.pixls.us/getfile.php/169/nice/Canon%20-%20EOS-1D%20X%20-%20RAW%20(3:2).CR2)
* [Pentax - K-1 - 14bit (3:2).DNG](https://raw.pixls.us/getfile.php/2470/nice/Pentax%20-%20K-1%20-%2014bit%20(3:2).DNG) (**Note:** This image requires the patch discussed in #765.)

One iPhone DNG found [here](https://www.austinmann.com/trek/iphone-14-pro-camera-review-scotland/) with BitsPerSample=**10**:
* [iPhone14Pro_ProRAW48MP.DNG](https://www.dropbox.com/s/kl764411n9jxr5c/iPhone14Pro_ProRAW48MP.DNG?dl=1) (**Note:** This image requires setting `cinfo.out_color_space = JCS_YCbCr` to be decoded properly.)

(The *.jpg files found in GoogleDrive were extracted from these files looking for SOI and EOI markers.)

I also added the image [14bit.pgm](https://drive.google.com/file/d/10lETW_bcLgbUdjccmeAli7XAVPh5lzo1/view?usp=drive_link) that I converted to Lossless JPEG using [libjpeg](https://github.com/thorfdbg/libjpeg):
```shell
jpeg -p -c -h 14bit.pgm 14bit.jpg
```
**Note:** `-h` is required or you will end up with a bad Huffman table (containing 256 codes).

The image generated by libjpeg can be converted back to PGM using libjpeg-turbo (with #769 applied):

```c
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include "jpeglib.h"

int main(int argc, const char * argv[]) {
    struct jpeg_decompress_struct cinfo;
    struct jpeg_error_mgr jerr;
    uint16_t *out;

    cinfo.err = jpeg_std_error(&jerr);
    jpeg_create_decompress(&cinfo);

    FILE *infile;

    static char *filename;
    filename = argv[1];
    if ((infile = fopen(filename, "rb")) == NULL) {
        fprintf(stderr, "can't open %s\n", filename);
        exit(1);
    }

    jpeg_stdio_src(&cinfo, infile);

    jpeg_read_header(&cinfo, TRUE);  // fails on data_precision without patch #769

    //cinfo.out_color_space = JCS_YCbCr;  // needed for Apple images

    int bit_depth = cinfo.data_precision;

    cinfo.data_precision = 16; // we want to use 16 bits to store the pixel values

    jpeg_start_decompress(&cinfo);

    cinfo.data_precision = bit_depth;  // needs to be correct when decoding

    size_t nbytes = 2 * cinfo.output_width * cinfo.output_height * cinfo.output_components;
    out = calloc(1, nbytes);

    while (cinfo.output_scanline < cinfo.output_height) {
        unsigned char *rowp[1];
        rowp[0] = (unsigned char *) out + cinfo.output_width * cinfo.output_components * cinfo.output_scanline * 2;
        jpeg16_read_scanlines(&cinfo, rowp, 1);
    }

    jpeg_finish_decompress(&cinfo);

    FILE *out_file = fopen("out.pgm", "wb");

    // store as PGM (pixel-interleaving the components)
    fprintf(out_file, "P5\n");
    fprintf(out_file, "%ld %ld\n", cinfo.output_width * cinfo.output_components, cinfo.output_height);
    fprintf(out_file, "%d\n", (1 << bit_depth) - 1);

    // byteswap to big-endian
    for (int i = 0; i < cinfo.output_width * cinfo.output_height * cinfo.output_components; i++) {
        out[i] = out[i] >> 8 | out[i] << 8;
    }

    jpeg_destroy_decompress(&cinfo);

    fwrite(out, nbytes, 1, out_file);

    fclose(out_file);

    return 0;
}

```

Now you can compare the PGM files:
```
> md5sum 14bit.pgm
584ea3ad82834300bf25b5167a376d2c  14bit.pgm
> md5sum out.pgm
584ea3ad82834300bf25b5167a376d2c  out.pgm
```

**Comment by SimonSegerblomRex** [comment]
> > Would it also be the case that, with a sample size of 8 or 12 bits, you can correctly decode any lossless JPEG image with that data precision or less?
>
> I think so, but I've only tried using jpeg16_read_scanlines so far.

I verified that setting `cinfo.data_precision = 12` and using `jpeg12_read_scanlines` produces exactly the same output for the iPhone 10 bit image (compared to using `jpeg16_read_scanlines`).

**Comment by SimonSegerblomRex** [comment]
> Hi @dcommander and @SimonSegerblomRex, This is very timely -- we recently found that MATLAB errors when reading a 14-bit lossless grayscale JPEG image in a DICOM file after I switched from using libjpeg v6b to using libjpeg-turbo v3.0.0. @SimonSegerblomRex: I see you've submitted a pull request in #769. Thank you for working on this! @dcommander: I know you won't be able to review the pull request until at least next week. I'll follow this issue to keep an eye on developments. Please reach out if you have any questions about our specific DICOM JPEG image.

Very interesting that you found a use case for DICOM as well @ellenjohnson! Is #759 enough for you? If possible it would be nice if you could upload an example image.

I think #759 is pretty much what's needed for the use case I had in mind (making libjpeg-turbo a viable option as a decoder for Lossless JPEG images in containers like DNG, CR2, and ARW for open-source RAW editors like Rawtherapee and Darktable), but I understand if @dcommander wants extend the scope a bit and fix more functions.

**Comment by dcommander** [comment]
Documentation is the main thing I'll need to add (in addition to your patch or a similar one) for 3.0.x, clarifying the difference between sample size and data precision (because the current implementation assumes that they are the same.) After I implement/integrate and document the decompressor modifications, I will look at adding the ability to generate lossless JPEG images with other data precisions as well. However, that will require more effort, and depending on how disruptive the compressor modifications are, they may have to wait until libjpeg-turbo 3.1.x.

I also need to develop a more thorough understanding of the interaction between sample size and data precision so I can make sure that libjpeg-turbo guards against decompressing a lossless JPEG image with data precision N into a packed pixel image with sample size < N (thus making the decompression process lossy.)

**Comment by ellenjohnson** [comment]
Thanks @dcommander for the update.  @SimonSegerblomRex - I can try a quick test to see if your changes allow us to decode the 14-bit JPEG image in our DICOM file.  As for uploading a sample image, I can't share the DICOM file but I'm checking if it's okay to share the embedded JPEG image which I've extracted.  Thanks again to both of you!

**Comment by SimonSegerblomRex** [comment]
> Here is a [very preliminary patch](https://github.com/user-attachments/files/15825253/0001-Handle-all-valid-lossless-data-precisions.patch) that allows decompressing lossless JPEG images with any data precision from 2-16 inclusive

Thanks! I just tested it quickly and it looks promising.

> , but unfortunately it still doesn't appear to work properly with the test images.
>
> 14bit.jpg The decompressed image appears too dark. [The other JPEG-1 reference implementation](https://github.com/thorfdbg/libjpeg) appears to decompress it correctly.

Your patch works just fine decoding this image! It's just maxval in the PGM header that gets the wrong value (65535 instead of 16383). The pixel values are correct:

```sh
> djpeg -outfile libjpeg-turbo.pgm 14bit.jpg
> tail -c +19 libjpeg-turbo.pgm | md5sum
9e5aa39da07911b7db132fd0abbe5ae5  -
> tail -c +19 14bit.pgm | md5sum
9e5aa39da07911b7db132fd0abbe5ae5  -

```

> Canon-EOS-1D-CR2-extracted.jpg, Pentax-K-1-DNG-extracted.jpg, Sony-ILCE-7RM5-ARW-extracted-tile.jpg: These images do not decompress even with [the other JPEG-1 reference implementation](https://github.com/thorfdbg/libjpeg). Are you sure they were extracted properly? libjpeg-turbo is unable to determine the colorspace of the images, which suggests a bad header.
>
> iPhone14Pro_ProRAW-DNG-extracted-tile.jpg This "lossless" JPEG image uses the YCbCr colorspace, so it isn't really lossless. libjpeg-turbo doesn't support lossy color conversions with lossless JPEG images. If I disable the relevant colorspace conversion check in jdcolor.c, it decompresses into a solid green output image, which doesn't seem correct. [The other reference implementation](https://github.com/thorfdbg/libjpeg) decompresses it into a pink and green image, which doesn't seem correct either.

I will have another look at these images, but generally there's not enough information in the JPEG header to know how to interpret the color components of Lossless JPEG images that are contained in DNG/TIFF or other RAW files. From the [DNG specification](https://helpx.adobe.com/content/dam/help/en/photoshop/pdf/DNG_Spec_1_7_1_0.pdf):
> For lossless JPEG, the internal width/length/components in the JPEG stream are not required to match the strip or tile's width/length/components. Only the total sample counts need to match. It is common for CFA images to be encoded with a different width, length or component count to allow the JPEG compression predictors to work across like colors. 

It's a bit tricky how to best handle this in `djpeg`. [libjpeg](https://github.com/thorfdbg/libjpeg) just saves each decoded component as a *.raw file when it doesn't know what to do. Personally I would be fine just pixel interleaving the components in a PGM file.

> I could also really use some examples of lossless JPEG images with 10-bit data precision and examples with < 8-bit data precision.

I might be able to find some 10 bit example images. <8 bit is probably hard to find in the wild, but I can generate them.

**Comment by dcommander** [comment]
> Your patch works just fine decoding this image! It's just maxval in the PGM header that gets the wrong value (65535 instead of 16383). The pixel values are correct:

Here's a [new patch](https://github.com/user-attachments/files/15828344/0001-Handle-all-valid-lossless-data-precisions.patch) that fixes that issue, so **14bit.jpg** now decompresses correctly at least.

Also, with the addition of this temporary patch:
```
--- a/jdcolor.c
+++ b/jdcolor.c
@@ -838,7 +838,9 @@ _jinit_color_deconverter(j_decompress_ptr cinfo)
   case JCS_EXT_BGRA:
   case JCS_EXT_ABGR:
   case JCS_EXT_ARGB:
-    if (cinfo->master->lossless && cinfo->jpeg_color_space != JCS_RGB)
+    if (cinfo->master->lossless && cinfo->jpeg_color_space != JCS_RGB &&
+        cinfo->jpeg_color_space != JCS_GRAYSCALE &&
+        cinfo->jpeg_color_space != JCS_YCbCr)
       ERREXIT(cinfo, JERR_CONVERSION_NOTIMPL);
     cinfo->out_color_components = rgb_pixelsize[cinfo->out_color_space];
     if (cinfo->jpeg_color_space == JCS_YCbCr) {
```

decompressing **iPhone14Pro_ProRAW-DNG-extracted-tile.jpg** now at least generates an output image with a similar shape to the one generated by [libjpeg](https://github.com/thorfdbg/libjpeg).  However, the colors are still different from those generated by [libjpeg](https://github.com/thorfdbg/libjpeg), and both output images still appear bogus.

When I implemented the lossless JPEG feature, I wrestled with whether libjpeg-turbo should decompress lossless JPEG images that were generated using RGB-to-YCbCr color conversion or chroma subsampling, since those are not reversible algorithms.  Since such images would never occur in medical imaging, I chose not to support them until I found examples of them in the wild.  However, such images are legal per the JPEG-1 spec, and apparently examples of them do exist in digital photography.  Thus, we should probably support decompressing (but not compressing) them ***if*** doing so is simply a matter of relaxing the colorspace check.  I would feel better about that if I understood what was going on with the colors in **iPhone14Pro_ProRAW-DNG-extracted-tile.jpg**.

Also, there is one potential pain point here, which is how to handle lossless JPEG images that don't have a JFIF marker or an Adobe marker that specifies the JPEG colorspace.  I'm not sure if the JPEG or JFIF specs have anything to say about such images, so if libjpeg-turbo cannot determine the JPEG colorspace from one of the aforementioned markers, then it always assumes the RGB colorspace for lossless JPEG images.  I'm not sure if that's technically correct, however, and it definitely isn't a correct assumption for lossless JPEG images compressed using `./jpeg -p -h` in [libjpeg](https://github.com/thorfdbg/libjpeg).  Such images use the YCbCr colorspace but contain neither a JFIF nor an Adobe marker.  (When adding an argument of `-c` to `./jpeg`, the lossless JPEG image will use the RGB colorspace and will contain an Adobe marker.)  Note that **iPhone14Pro_ProRAW-DNG-extracted-tile.jpg** contains a JFIF marker, so this paragraph doesn't apply to the issue described above.

libjpeg-turbo should also support decompressing grayscale lossless JPEG images into RGB and extended RGB packed-pixel images, since that is a mathematically lossless color conversion (nothing controversial there-- just an oversight on my part.)

> It's a bit tricky how to best handle this in `djpeg`. [libjpeg](https://github.com/thorfdbg/libjpeg) just saves each decoded component as a *.raw file when it doesn't know what to do. Personally I would be fine just pixel interleaving the components in a PGM file.

The calling program would have to be responsible for that, since there is no such facility in libjpeg-turbo.

> I might be able to find some 10 bit example images. <8 bit is probably hard to find in the wild, but I can generate them.

Can you tell me how to generate them?  That would probably be sufficient for now.  Ideally, I would like to create an automated script that generates lossless JPEG images using [libjpeg](https://github.com/thorfdbg/libjpeg) with all bit depths from 2-16, then compares the decompressed results between [libjpeg](https://github.com/thorfdbg/libjpeg) and libjpeg-turbo.

The funding for this work comes from MathWorks, who prepaid for a certain number of hours of maintenance and bug fixing on the lossless JPEG feature.  Thus, the primary focus at the moment is how to support DICOM lossless JPEG images with data precisions other than 8, 12, or 16.  If I can do some additional logic twiddling to support certain DNG images, such as those that use the YCbCr colorspace, then I am open to that.  But anything that involves more than logic twiddling is probably out of scope or would minimally have to be pushed to libjpeg-turbo 3.1.  (I'm already nervous enough about introducing this into a bug-fix release, since inevitably even the most minor and seemingly innocuous changes in this code manage to break something somewhere.)

**Comment by dcommander** [comment]
I figured out how to reduce the colors in a given PPM file by using the `+dither` and `-depth` arguments in ImageMagick's `convert` program, so I can now use [libjpeg](https://github.com/thorfdbg/libjpeg) to create lossless JPEG images with any data precision from 2 to 16.  There definitely does still appear to be an issue with how libjpeg-turbo handles certain data precisions, so I am investigating that.

**Comment by dcommander** [comment]
Here is the [latest patch](https://github.com/user-attachments/files/15841639/0001-Handle-all-valid-lossless-data-precisions.patch).  It seems to work properly for 2 to 16-bit RGB or grayscale lossless JPEG images when using djpeg, although I still need to do more automated testing, tweak the documentation, and make sure everything works properly with the TurboJPEG API.  (More unit tests may be required.)

Because this feature expands our coverage of the JPEG spec beyond that which is claimed by libjpeg-turbo 3.0.x, and because expanding that coverage potentially introduces new attack vectors, I am leaning toward pushing it into the dev branch (thus deferring the release of the feature until libjpeg-turbo 3.1 beta.)  I feel that that is the most prudent approach for the upstream project, although downstream integrators will be free to back-port the patch into their implementations of libjpeg-turbo 3.0.x as they see fit.  I can also look at adding compression support for 2 to 16-bit lossless JPEG images in libjpeg-turbo 3.1.x, as well as support for colorspace conversions (per above.)  I feel that putting all of that into dev would give the feature a chance to evolve and would allow for downstream testing/vetting of the approach before it is officially released.

**Comment by dcommander** [comment]
The other thing to bear in mind is that my funding is quickly running out.  I think I can finish the decompression support for 2 to 16-bit lossless images using existing funding, but compression support may have to wait a while.  That's potentially a problem vis-a-vis vetting the feature, because I need compression support in order to create proper unit tests.

**Comment by dcommander** [comment]
Never mind about funding. Microsoft just dropped me some cash out of the blue, so I can use it for investigating/adding compression support for 2-16 bit lossless images.

**Comment by SimonSegerblomRex** [comment]
> Because this feature expands our coverage of the JPEG spec beyond that which is claimed by libjpeg-turbo 3.0.x, and because expanding that coverage potentially introduces new attack vectors, I am leaning toward pushing it into the dev branch (thus deferring the release of the feature until libjpeg-turbo 3.1 beta.) I feel that that is the most prudent approach for the upstream project, although downstream integrators will be free to back-port the patch into their implementations of libjpeg-turbo 3.0.x as they see fit. I can also look at adding compression support for 2 to 16-bit lossless JPEG images in libjpeg-turbo 3.1.x, as well as support for colorspace conversions (per above.) I feel that putting all of that into dev would give the feature a chance to evolve and would allow for downstream testing/vetting of the approach before it is officially released.

Sounds like a good strategy!

> Also, there is one potential pain point here, which is how to handle lossless JPEG images that don't have a JFIF marker or an Adobe marker that specifies the JPEG colorspace.

I ran [Pentax - K-1 - 14bit (3:2).DNG](https://raw.pixls.us/getfile.php/2470/nice/Pentax%20-%20K-1%20-%2014bit%20(3:2).DNG) through the latest release of the Adobe DNG Converter and extracted this tile: ![adobe-dng-converter-extracted-tile](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/2533963d-c7f5-4fd6-a455-eac8de1c220f)

It doesn't contain any information about the color space either (probably since it doesn't really make sense for [CFA](https://en.wikipedia.org/wiki/Color_filter_array) images anyway) , so I think it's important that the library supports decoding these images, but it's OK if `djpeg` doesn't. (Using 2 components in combination with predictor mode 1 is very common when encoding CFA images.) Perhaps the best solution for `djpeg` would be to assume grayscale and encode the output as PGM.

**Comment by ellenjohnson** [comment]
Thanks @dcommander for the patch!   I'll apply the patch to our libjpeg-turbo v3.0.0 and integrate it with our MATLAB code to verify if we MATLAB can now read our 14-bit JPEG.  
Meanwhile, I'm attaching the 14-bit lossless JPEG I extracted from the DICOM file.  Can you verify if you can decode this image with after your changes?  Thanks!

![dcm1](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/50502729/69a8e195-0d3d-47e7-9b54-73fa2d2214eb)

**Comment by dcommander** [comment]
@ellenjohnson Do **not** apply the patch.  It has known issues.  Please wait until I can resolve this the right way.  I am about to post a long comment about that.

**Comment by dcommander** [comment]
The decompression side of this feature has been stabilized and extended to the TurboJPEG API. It passes a rigorous automated test script that uses ImageMagick to generate 2-bit through 7-bit source images from the 8-bit test images available [here](http://www.imagecompression.info/test_images) and 9-bit through 15-bit source images from the 16-bit test images available [here](http://www.imagecompression.info/test_images), uses [libjpeg](https://github.com/thorfdbg/libjpeg) to generate lossless JPEG images from the source images, and compares the decompressed results from  djpeg and TJBench (C and Java) with the source images.  It is now more caller-proof than the patches I posted above (which would cause a segfault if you, for instance, tried to call one of the `jpeg*_read_scanlines()` functions with the wrong data precision.)

The more I dig into it, however, the more I realize that it doesn't make sense to push the feature until I can implement the compression side as well. Even though this feature requires no API or other interface changes, it needs to be treated with the same rigor with which I treated 16-bit lossless support. That means adding appropriate unit tests. As I work on the compression side, I may find additional issues with the decompression side.  (I have already refined the implementation somewhat relative to the rough patches I posted above.)  Also, the nature of the documentation is such that it would be very awkward and time consuming to document the feature as a decompression-only feature.

The feature definitely will be pushed to the dev/3.1 evolving branch and not back-ported into 3.0.x. The fact that my first attempt at it reduced the robustness of the libjpeg API, potentially exposing the library to a segfault (potential security issue) if the caller made a simple mistake, tells me that I'm playing with too much fire to justify including this in a bug-fix release.  The good news is that 3.1 shouldn't take years to land.  This feature, #645, and #771 are probably enough to justify it, so hopefully it can go into beta later this year.

I should be able to finish the compression side and push the complete feature sometime next week.  Please wait on attempting to integrate the feature downstream until it is pushed.

Now, as to the colorspace issues:

I have decided to stick to my guns and not allow color conversion or upsampling/downsampling with lossless JPEG images.  Philosophically, I don't think it makes sense for the library to allow any decompressed output from a lossless JPEG image that is not mathematically lossless relative to the original image.  Calling programs will still be free to decompress a lossless YCbCr JPEG image into a packed-pixel YCbCr image, and I also intend to investigate whether I can open up the planar ("raw" in the libjpeg API, "YUV" in the TurboJPEG API) compression/decompression interfaces for lossless images.  (I mainly avoided doing that in libjpeg-turbo 3.0.x due to lack of time to test it.)  Apart from that, it really doesn't make sense to use lossless JPEG if you intend to use an irreversible color conversion.  You'd be better off using DCT-based JPEG with quality 100.

I am seeking clarification from Thomas as to whether YCbCr is specified as the default colorspace for JPEG images or whether that's an implementation detail, but that isn't relevant for any of the DNG test images above.  The only one that appears to use YCbCr is the iPhone ProRAW image, and it has a JFIF marker to indicate that.  However, even if I loosen the colorspace restrictions for lossless JPEG images, libjpeg-turbo still doesn't decompress that image properly, nor does [libjpeg](https://github.com/thorfdbg/libjpeg).  (It's supposed to be an image of a highland cow in Scotland.  Incidentally, it was taken by the husband of a photographer I have worked with in the past.)

> It doesn't contain any information about the color space either (probably since it doesn't really make sense for [CFA](https://en.wikipedia.org/wiki/Color_filter_array) images anyway) , so I think it's important that the library supports decoding these images, but it's OK if `djpeg` doesn't. **EDIT:** With this patch `djpeg` produces the output I expect:

Yeah, of course libjpeg-turbo won't decode that image anyhow, since the image uses an invalid Huffman code.  Irrespective of that, the image has 2 components, and the IJG's software has never been able to handle 2-component JPEG images.  There is simply not enough information to figure out what the components are supposed to be.  Nothing in the library prevents calling programs from handling such images as they see fit.  (The fact that you were able to patch djpeg to make it handle the image demonstrates that.)  However, djpeg certainly cannot assume that any 2-component image with an unknown JPEG colorspace is CFA.

**Comment by dcommander** [comment]
Removed the preliminary patches from my comments above, lest anyone try to integrate them.  (They were meant for bench testing only, not for production.)  I will let you know when the full feature has been pushed into the dev branch.  I have other commitments this week, so it will probably be next week before I can get to it.  Please be patient.  Hacks are quick, but making something production-ready in a library that is used by billions of people every day is slow.

**Comment by dcommander** [comment]
@ellenjohnson Yes, the latest code I am working on does appear to correctly decode that image.

**Comment by SimonSegerblomRex** [comment]
Thanks for the update @dcommander!

> Yeah, of course libjpeg-turbo won't decode that image anyhow, since the image uses an invalid Huffman code. Irrespective of that, the image has 2 components, and the IJG's software has never been able to handle 2-component JPEG images.

Actually this was the output from Adobe DNG Converter, so at least the Huffman codes are valid.

> However, djpeg certainly cannot assume that any 2-component image with an unknown JPEG colorspace is CFA.

I agree that there's no correct way for how `djpeg` should handle 2-component images (or images with more components with unknown color information) and it's perfectly OK not to support this case. (The most helpful thing for the user would probably be to pixel-interleave the components in a PGM though, but let's not include that in this feature request.)

**Comment by dcommander** [comment]
Sounds good.  Once the feature is pushed, you can test it and tell me if there are additional cases that the library may need to handle.

**Comment by SimonSegerblomRex** [comment]
> Thanks @dcommander for the patch! I'll apply the patch to our libjpeg-turbo v3.0.0 and integrate it with our MATLAB code to verify if we MATLAB can now read our 14-bit JPEG. Meanwhile, I'm attaching the 14-bit lossless JPEG I extracted from the DICOM file. Can you verify if you can decode this image with after your changes? Thanks!
> 
> ![dcm1](https://private-user-images.githubusercontent.com/50502729/341123155-69a8e195-0d3d-47e7-9b54-73fa2d2214eb.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4MTA5OTIsIm5iZiI6MTcxODgxMDY5MiwicGF0aCI6Ii81MDUwMjcyOS8zNDExMjMxNTUtNjlhOGUxOTUtMGQzZC00N2U3LTliNTQtNzNmYTJkMjIxNGViLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDE1MjQ1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVlMzEwMTgyNWYyODk3MmQ2YmI0YjllODNlMDFkNjdhNjA3MWEzYzcxZjEzNDA5ZDFjYzVhYTMxZTZiOTAxNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XI2Rut-AabjNqvlIaOHV7_JEoo5oC3zLYRDtapgvHdI)

@ellenjohnson I just did a quick test decoding this image with the last uploaded WIP patch and `djpeg` decodes the images just fine (bit exact output compared with `jpeg` from [libjpeg](https://github.com/thorfdbg/libjpeg)).

**Comment by ellenjohnson** [comment]
Thanks @dcommander and @SimonSegerblomRex for testing the 14-bit JPEG I shared -- I really appreciate it!
@dcommander I will hold off doing further integration of the patch until you release the full libjpeg-turbo feature as you described.  Thanks again for working on this!
Meanwhile, we got new customer report (from different customer) that their 10-bit DICOM file cannot be read with MATLAB after I switched from libjpeg v6b to libjpeg-turbo v3.0.0.  I reached out asking permission if I can share the embedded JPEG.  I'll give an update if I get permission.

**Comment by dcommander** [comment]
I am confident that it will work.  DICOM appears to use the more common RGB-colorspace lossless JPEG variant, as opposed to the unusual colorspaces used by the DNG files mentioned above.

**Comment by dcommander** [comment]
Compression is fully working and passing the regression tests. So far I am about 30 hours into this, and I still need to finish documenting the feature and adding unit tests. I have already burned through what was left of MathWorks' funding and am at risk of also burning through the funding I just got from Microsoft. Any additional financial assistance that MathWorks can provide toward this end would be appreciated.

**Comment by dcommander** [comment]
A proper implementation of this has been checked into the dev (3.1 evolving) branch.  You can see from that commit how much more effort was required (including documentation and testing and code robustness) in order to fully support these new data precisions.  The implementation also uncovered some other issues and shortcomings that I had to fix.  (Refer to recent commits in main and dev.)  Total time for all of the above was 68 hours, with probably about 50 hours of that attributable specifically to this feature.  Such is the maddening nature of this project.  If you want to change anything in the libjpeg API and associated programs, you find about ten other things that you need to change first.  That's why, after 14 years, I have a strong revulsion toward making any major changes to the libjpeg API and its associated programs.  That API really needs to be redesigned from the ground up (see #313 and #646 for my general musings) before any other major changes are made to it.  However, I also have a strong feeling that no one really cares enough about this old image format to pay for such sweeping changes.  Hopefully this is at least the last major coverage change, since JPEG-1 doesn't have any other modes to cover (except hierarchical, which we will never support.)

I wish I would have known about the need to support these additional data precisions prior to implementing 16-bit data precision.  I would have condensed the sample types and function signatures into two (one for 2 to 8 bits and one for 9 to 16 bits) instead of three.  (The use of a signed short for `J12SAMPLE` is a historical relic, but as soon as I changed the type name, I didn't actually have to conform to that legacy.)  So it goes.  Practically every major feature in this project has been, in some way, an effort to work around someone's (including my own) inability to predict the future.

**Comment by SimonSegerblomRex** [comment]
Thank you @dcommander! I'll review the changes and start testing.

> I wish I would have known about the need to support these additional data precisions prior to implementing 16-bit data precision. I would have condensed the sample types and function signatures into two (one for 2 to 8 bits and one for 9 to 16 bits) instead of three. (The use of a signed short for J12SAMPLE is a historical relic, but as soon as I changed the type name, I didn't actually have to conform to that legacy.)

I guess this is too late now, but from a usability point of view it would have been nice to let `jpeg16_read_scanlines` handle all valid bit depths (at least above 8). Not a big thing, but being forced to use `jpeg12_read_scanlines` for the 10 bit use-case makes things a bit messier for applications using the library.

**Comment by dcommander** [comment]
Yeah, this is the difference between making something work for a particular use case and making it work for all use cases.  The design of the libjpeg API never took into account lossless JPEG, nor did it take into account the need to handle multiple data precisions at run time.  libjpeg was meant to be configured for a particular use case using compile-time macros, so you could use those macros to disable unused features or change the RGB component order or change the sample data type or change the data precision, etc.  The concept of shared libraries with fixed APIs came later, and once Linux distributors started shipping fixed builds of libjpeg, the default implementation of the libjpeg API became a sort of pseudo-standard.  That default API implementation only supports 8-bit data precision, uses `unsigned char` for samples, and uses a fixed RGB component order.  The first thing I had to do in libjpeg-turbo was to introduce colorspace extensions, which allow multiple RGB component orders to be selected at run time.  Then I introduced a CMake variable that at least made it possible to build the library with 12-bit data precision without modifying the headers.  For 13 years, that was the only way you could use libjpeg-turbo with a data precision other than 8-bit.  No O/S ever shipped a 12-bit version of the library, in part because building it with 12-bit data precision didn't change any of the function/data type names.  Applications that attempted to support both 8-bit and 12-bit data precision had to include separate in-tree builds of libjpeg-turbo and deal with the resulting namespace collisions.  Thus, few applications supported data precisions other than 8-bit.

Run-time-selectable data precision and lossless JPEG support in libjpeg-turbo is less than a year old and wouldn't have been possible without generous funding from MathWorks.  I added run-time-selectable 12-bit data precision first, and when I added that feature, I chose to retain backward compatibility with the previous compile-time-selectable 12-bit implementation.  Thus, the 12-bit interfaces/data types are basically just renamed versions of the interfaces/data types you got with libjpeg-turbo 2.1.x and prior when you set the `WITH_12BIT` CMake variable to `1`.  There was an advantage to that, because it allowed applications that had previously used the compile-time-selectable 12-bit implementation to simply change `JSAMPLE` to `J12SAMPLE` and `jpeg_read_scanlines()`/`jpeg_write_scanlines()` to `jpeg12_read_scanlines()`/`jpeg12_write_scanlines()`.  When I added 16-bit lossless support, which was still before libjpeg-turbo 3.0 beta was released, I had the opportunity to use `J16SAMPLE` for all data precisions higher than 8-bit.  In retrospect, I should have done that.  However, you have to remember that the lossy JPEG algorithms have a fixed sample size and can't bit shift like the lossless JPEG algorithms can.  Also, the 8-bit lossy algorithms are SIMD-accelerated, so they absolutely must use byte samples.  Thus, it isn't possible to decompress an 8-bit lossy JPEG image using the 12-bit or 16-bit functions, and it never will be.

The libjpeg architecture also wouldn't easily support decompressing the same lossless JPEG image into 8-bit or 16-bit samples.  In order to support multiple data precisions at run time, each module has to be compiled multiple times with different compile-time values of `BITS_IN_JSAMPLE`.  (See above regarding the fact that run-time-selectable data precision was never taken into account when libjpeg was designed.)  Thus, there is an 8-bit version of the modules, a 12-bit version of the modules, and a 16-bit version of the modules.  The library has to use an 8-bit, 12-bit, or 16-bit data path throughout the compression/decompression process, so only one set of modules can be enabled simultaneously.  Those modules have to be enabled in the body of `jpeg_start_compress()`/`jpeg_start_decompress()`, so at that point, the library has to agree with the caller which data path to use.  The only way to reach that agreement is to choose a single data path based on the value of `cinfo->data_precision`.  In order for the 12-bit and 16-bit functions to handle any data precision <= `BITS_IN_JSAMPLE`, I would have to make the following deep modifications:

- If we're compressing or decompressing a lossless JPEG image, initialize all modules in all data paths, regardless of the value of `cinfo->data_precision`.  (It is documented that the `data_precision` field must be initialized prior to calling `jpeg_start_compress()`/`jpeg_start_decompress()`, so the hack you did above is not a legal use of the API.)
- Introduce a sample size field in the opaque `jpeg_comp_master` and `jpeg_decomp_master` structs.
- In the body of `jpeg*_read_scanlines()` and `jpeg*_write_scanlines()`, set the sample size field to the value of `BITS_IN_JSAMPLE` if we're compressing or decompressing a lossless JPEG image. (In other words, `jpeg12_read_scanlines()` would set the field to 12, and `jpeg16_read_scanlines()` would set the field to 16.)
- If we're compressing or decompressing a lossless JPEG image, select the data path based on the sample size field rather than `cinfo->data_precision`.
- Figure out how to document and test all of the above, which would not be straightforward.

However, given that the data precision affects everything down to the memory manager level, I can foresee about 100 issues that might arise from this.  Also, it wouldn't be the common case.  (Most people don't want to intentionally use more memory than necessary when using a compression library.)  Thus, I would have to formulate new unit tests for it.  After having just spent 50+ hours testing and documenting this mess, I can tell you that what you're asking for would be a lot messier to test and document-- not to mention potentially confusing to users. (Remember what I said above: hacks are quick, but making something production-ready in a library that is used by billions of people every day is slow.)  I am out of funding.  Finished.  Kaput.  I had a nice surplus in the libjpeg-turbo General Fund, and this feature completely wiped it out.  So at this point, any further modifications will have to be simple ones that don't require a massive amount of re-documenting and re-testing.  Our PPM I/O modules demonstrate how to bit-shift decompressed output from 2-bit to 15-bit lossless JPEG images into a 16-bit buffer.

There is still the opportunity to revisit this before libjpeg-turbo 3.1 beta is released, but again, any change in behavior would have to conform to the previously documented API behavior and would require a great deal of additional labor for re-documentation and re-testing.  I would have to receive retroactive funding for the feature as implemented, as well as additional funding for the aforementioned modifications.  If that funding emerges, it will likely emerge from a company that needs to support 2-16-bit lossless JPEG images in the context of medical imaging.  If that company doesn't need to compress/decompress lossless JPEG images with 2-8 bits of data precision from/to 16-bit-wide samples, then there won't be funding to implement that feature.

**Comment by dcommander** [comment]
To comment further on the memory manager issues, the `alloc_sarray()` and `access_virt_sarray()` methods return a `JSAMPLE`, `J12SAMPLE`, or `J16SAMPLE` depending on the value of `cinfo->data_precision`, and there are some esoteric modes in which applications may need to directly allocate memory or do other things that would break if we didn't fix the data path in the body of `jpeg_start_compress()`/`jpeg_start_decompress()`.  I strongly suspect that, if I were able to make it work as you request, it would only work for a very limited set of use cases (which means that I would have to carefully test and document those use cases.)

**Comment by SimonSegerblomRex** [comment]
> Also, it wouldn't be the common case.

I agree, letting `jpeg16_read_scanlines` support bit-depths <= 8 is not a use-case I see much value in supporting, and it's definitely not worth the effort. Bit-depths 9–16 would be nice, but it's easy enough for the calling application to add an if-statement combining `jpeg12_read_scanlines` and `jpeg16_read_scanlines`. For the applications I have in mind the only crucial bit-depths to support are 10, 12, 14, and 16.

Except for the issue discussed in #765 I think all functionality needed for RAW imaging applications is in place now. (I found [some more images](https://github.com/cgohlke/imagecodecs/pull/105/files#r1652591886) with invalid Huffman codes, but let's follow that up in #765.)

It would be nice with some input from @adobe as well. Last time I checked they were using [jpeg-9c](http://www.ijg.org/files/jpegsr9c.zip) for lossy JPEG and their own version of Thomas Lane's Lossless JPEG implementation.

**Comment by SimonSegerblomRex** [comment]
I've verified that I can decode the following images (from [here](https://drive.google.com/drive/folders/1ocDXK1bvWSOOZ66Wwp-0pEVhxobLVf2i?usp=sharing)) correctly (without any modifications or using the library in an unsupported way):
* Canon-EOS-1D-CR2-extracted.jpg
* Sony-ILCE-7RM5-ARW-extracted-tile.jpg
* iPhone14Pro_ProRAW-DNG-extracted-tile.jpg (**Note:** Requires specifying the colorspace before calling `jpeg_start_decompress`.)

Using [imagecodecs](https://github.com/cgohlke/imagecodecs/pull/105) libjpeg-turbo offers ~40 % faster decoding compared to the old Lossess JPEG library lj92. :rocket:

**Comment by dcommander** [comment]
> I agree, letting `jpeg16_read_scanlines` support bit-depths <= 8 is not a use-case I see much value in supporting, and it's definitely not worth the effort. Bit-depths 9–16 would be nice, but it's easy enough for the calling application to add an if-statement combining `jpeg12_read_scanlines` and `jpeg16_read_scanlines`. For the applications I have in mind the only crucial bit-depths to support are 10, 12, 14, and 16.

I agree that it makes more sense to attach 9-bit through 11-bit and 13-bit through 16-bit data precision to the 16-bit functions, since those functions can only be used with lossless mode.  The question would be how to handle 12-bit data precision.  Possibilities:

1. Leave 12-bit data precision attached only to the 12-bit functions, and attach the other data precisions > 8-bit to the 16-bit functions.  That could be done without the aforementioned deep modifications to the library. However, it would be almost as awkward as what we're currently doing, since a calling application would still have to use an `if()` statement.
2. Use the 12-bit functions only for lossy mode, and use the 16-bit functions for lossless mode with 9-bit through 16-bit (including 12-bit) data precision.  That could also be done without the aforementioned deep modifications to the library, but it would be potentially confusing to users and hard to document.  It would also represent an API backward incompatibility, since the 12-bit functions in libjpeg-turbo 3.0.x handled both lossy and lossless mode.  The API backward incompatibility makes this option a non-starter.
3. Leave 12-bit data precision attached only to the 12-bit functions, and attach all data precisions > 8-bit (including 12-bit) to the 16-bit functions. I like this approach the best.  It's the least confusing and easiest to document, but handling the overlap between the two function families for 12-bit data precision would be painful. It wouldn't be as painful as handling data precisions <= 8-bit with the 16-bit functions, but it would have some significant percentage of that pain.  It would still be necessary to preemptively initialize both the 12-bit and 16-bit data paths in lossless mode, `jpeg(12|16)_read_scanlines()` and `jpeg(12|16)_write_scanlines()` would still need to record which data path the calling application is requesting, and the lower-level functions would still need to act upon that data path variable rather than `cinfo->data_precision`.  The chance of breakage is less than if we were trying to handle single-byte data precisions with the 16-bit functions, because at least `J12SAMPLE` and `J16SAMPLE` are the same size.  However, because cjpeg, djpeg, and the TurboJPEG API have to handle lossy and lossless mode, they would still call the 12-bit functions for 12-bit data precision.  Thus, the 16-bit functions wouldn't receive a lot of testing with 12-bit data precision.

All of that is a moot point without additional funding.

> Except for the issue discussed in #765 I think all functionality needed for RAW imaging applications is in place now.
> 
> It would be nice with some input from @adobe as well. Last time I checked they were using [jpeg-9c](http://www.ijg.org/files/jpegsr9c.zip) for lossy JPEG and their own version of Thomas Lane's Lossless JPEG implementation.

Oof.  No one should really be using jpeg-8* or jpeg-9* if they care about compatibility with the spec.  In jpeg-8, Guido introduced an unsanctioned extension to the JPEG-1 format ("SmartScale"), which allows NxN (N=1..16) DCT block sizes.  In jpeg-9, he introduced a reversible color transform that is also, to the best of my understanding, not conformant with the spec.  The combination of a 1x1 DCT block size and the reversible color transform allows him to generate lossless JPEG-like images using the DCT algorithm, but those images are not actually JPEG images.  When the ISO and ITU-T rejected his unsanctioned extensions, he started using the IJG README file as a manifesto against them (and me.)  The IJG's software is still developed behind closed doors (no public repository) and doesn't receive the same security testing that libjpeg-turbo receives.

Tom Lane never had a lossless JPEG implementation.  He stepped back from the IJG in 1998, and the lossless JPEG implementation was released a year later by Ken Murchison as a patch against jpeg-6b.  Ken's patch (with heavy modifications) is the basis for the implementation in libjpeg-turbo.  If you're referring to https://github.com/thorfdbg/libjpeg, that is Thomas Richter's implementation (which is, confusingly, also called "libjpeg.")  
Thomas Richter is a member of the JPEG working group, and both his software and libjpeg-turbo are JPEG-1 reference implementations.

**Comment by SimonSegerblomRex** [comment]
> Leave 12-bit data precision attached only to the 12-bit functions, and attach all data precisions > 8-bit (including 12-bit) to the 16-bit functions. I like this approach the best.

I agree that this sounds like the best option! But things are good enough as they are now :+1:

> Tom Lane never had a lossless JPEG implementation. He stepped back from the IJG in 1998, and the lossless JPEG implementation was released a year later by Ken Murchison as a patch against jpeg-6b. Ken's patch (with heavy modifications) is the basis for the implementation in libjpeg-turbo. If you're referring to https://github.com/thorfdbg/libjpeg, that is Thomas Richter's implementation (which is, confusingly, also called "libjpeg.")
Thomas Richter is a member of the JPEG working group, and both his software and libjpeg-turbo are JPEG-1 reference implementations.

Right, I confused the names, then Adobe must have based it on something else. (They refer to
```
// Lossless JPEG code adapted from:

/* Copyright (C) 1991, 1992, Thomas G. Lane.
 * Part of the Independent JPEG Group's software.
 * See the file Copyright for more details.
 *
 * Copyright (c) 1993 Brian C. Smith, The Regents of the University
 * of California
 * All rights reserved.
 * 
 * Copyright (c) 1994 Kongji Huang and Brian C. Smith.
 * Cornell University
 * All rights reserved.
```
)

I found the numbers Adobe that boasts with here a bit amusing:
![image](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/17ad34ce-1e95-4d1d-bf46-7685ae2ad4a2)
https://opensource.adobe.com/

**Comment by dcommander** [comment]
> Right, I confused the names, then Adobe must have based it on something else. (They refer to
> 
> ```
> // Lossless JPEG code adapted from:
> 
> /* Copyright (C) 1991, 1992, Thomas G. Lane.
>  * Part of the Independent JPEG Group's software.
>  * See the file Copyright for more details.
>  *
>  * Copyright (c) 1993 Brian C. Smith, The Regents of the University
>  * of California
>  * All rights reserved.
>  * 
>  * Copyright (c) 1994 Kongji Huang and Brian C. Smith.
>  * Cornell University
>  * All rights reserved.
> ```

That is the "Cornell codec" mentioned in various comments under https://github.com/libjpeg-turbo/libjpeg-turbo/issues/402.

**Comment by SimonSegerblomRex** [comment]
Just adding this info for future reference:

I don't have any experience working with DICOM files, but out of curiosity I checked if libjpeg-turbo could decode the Lossless JPEG images embedded in [this data set](https://sourceforge.net/projects/gdcm/files/gdcmData/).

From the data set I managed to extract the following (just looking for SOI + SOF3 + EOI markers):

| Bit-depth | Number of images |
| --------: | ---------------: |
|         8 |               86 |
|        10 |                3 |
|        12 |                6 |
|        15 |                2 |
|        16 |               16 |

The images with bit-depth 15 are a bit extra interesting since I haven't seen @ellenjohnson request support for this bit-depth:

| Filename | Extracted Lossless JPEG | `gdcmraw` output |
| :------- | :--------------------: | :--------------: |
| D_CLUNIE_RG1_JPLL.dcm | ![D_CLUNIE_RG1_JPLL-extracted](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/3fa29775-96f5-44d1-baa7-25d3996c3cef) | ![D_CLUNIE_RG1_JPLL-gdcmraw-output](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/53f792a0-0abd-4887-b513-f1dda3707a3d) |
| gdcm-JPEG-LossLessThoravision.dcm | ![gdcm-JPEG-LossLessThoravision-extracted](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/66872a6d-24bf-46e2-9599-bfd8442365dc) | ![gdcm-JPEG-LossLessThoravision-gdcmraw-output](https://github.com/libjpeg-turbo/libjpeg-turbo/assets/8997425/ff0bf644-c14b-45c5-adbf-38e2afb2b3b5) | 

libjpeg-turbo can decode the output from `gdcmraw` without problem, **but** the extracted files (not processed with gdmcraw) are broken:

| Filename | `jpeg` (from [thorfdbg/libjpeg](https://github.com/thorfdbg/libjpeg)) | `djpeg` (libjpeg-turbo) |
| :--------| :-------------------------------------------------------------------- | :---------------------- |
| D_CLUNIE_RG1_JPLL-extracted.jpg | WARNINGS: <br> `Reason is: expecting a marker or marker segment - stream is out of sync` <br> `Reason is: found invalid marker, probably a marker size is out of range` | ERROR: <br> `Corrupt JPEG data: 457 extraneous bytes before marker 0xd9` |
| gdcm-JPEG-LossLessThoravision-extracted.jpg | ERROR: <br> `invalid stream, found invalid huffman code in entropy coded segment` | ERROR: `Corrupt JPEG data: bad Huffman code` |

This must be due to some bug in the encoder that encoded the images and they are not actually valid JPEG images, so I'm not arguing that libjpeg-turbo does anything wrong. If you get any support requests originating from broken DICOM files then the best thing to do is probably to refer to `gdcmraw` as a repair tool.

**Comment by dcommander** [comment]
That's good info to have.  #402 discusses the various issues that have been encountered with DICOM lossless JPEG images "in the wild" and the options that DCMTK (with uses libjpeg + Ken Murchison's patch) added in order to work around them.  (The aforementioned Cornell encoder was apparently notorious.)  Tom Lane created the IJG and released jpeg-1 before the JPEG-1 and JFIF specs were even released, and the availability of a liberally-licensed open source codec that implemented a particular lossy variant of JPEG-1 encouraged convergence around that variant.  Such did not happen with lossless JPEG-1, however, so there was a bit of a "Wild West" scenario in the 1990s in which different lossless codecs emerged with slightly different implementations (some of them incorrect, such as the Cornell codec.)  I suspect that the `--workaround-incpl` option in DCMTK was meant to support images such as the ones you found.

**Comment by dcommander** [comment]
Upon further examination, the 8-bit, 12-bit, and 16-bit data paths are mutually exclusive, because they allocate precision-specific controller structures and link the `jpeg_compress_struct` or `jpeg_decompress_struct` instance to those structures.  (That occurs within the body of `jpeg_start_compress()`/`jpeg_start_decompress()`.)  Thus, the "ideal" option (Option 3 above) isn't possible without modifying the library in a very deep way that, frankly, isn't worth the effort.  That leaves only Option 1 as an alternative, and I don't like that option as much as what we're already doing.  What we're already doing is consistent, since it extends each of the data-precision-specific functions to handle data precisions less than the maximum data precision supported by that function (but not less than or equal to the maximum data precision supported by another function.)  I honestly think that the only additional work that needs to be done on this feature is to extend the example programs to handle lossless JPEG images with 2-16 bits of data precision.  (The examples don't currently handle lossless JPEG images at all.)

**Comment by SimonSegerblomRex** [comment]
> I honestly think that the only additional work that needs to be done on this feature is to extend the example programs to handle lossless JPEG images with 2-16 bits of data precision.

I agree! One interesting aspect (that probably doesn't need to be covered in the documentation, but that still is good to keep in mind) is that the value of bits per sample ("P" in the scan header, only used as an exponent calculating the first prediction value) doesn't affect the compression ratio for Lossless JPEG. So if you have a 12 bit image it doesn't really matter if you store it using bits per sample 12, 13, 14, 15, or 16 from a compression point of view. The decoder has to use the same value for bits per sample that was used in the encoding process. For further processing of the decoded image it's still important to know the "correct" value of bits per sample (e.g., to get the right level or know how many bits per sample you should use when converting the file to some other image format), so you still have every reason to use the correct value when encoding the original image. However, in the DNG use-case you can specify the WhiteLevel tag to know the real bit depth of the pixel values, and then it doesn't matter if you used a bogus value for the embedded Lossless JPEG encoded image. Indeed this is what Adobe seems to do: If you run a DNG file with an embedded 14 bits per sample Lossless JPEG image through the Adobe DNG Converter it will be decoded (using 14 bits per sample) and then encoded using 16 bits per sample for the Lossless JPEG (without touching the original pixel values, they are still 14 bit values).

**Comment by dcommander** [comment]
How does that interact with the point transform? If you have, say, 16-bit-wide samples with 14-bit data precision, would it be approximately equivalent to encode them as a 14-bit lossless JPEG vs. a 16-bit lossless JPEG with a Pt of 2?

**Comment by dcommander** [comment]
If I recall how the Pt works, it is basically a right shift of the input samples, so what I'm asking is whether it would be roughly equivalent to compress:
- 16-bit-wide samples, with only the right-most 14 bits used, as a 14-bit lossless JPEG image and
- 16-bit-wide samples, with only the left-most 14 bits used, as a 16-bit lossless JPEG image with Pt=2.

**Comment by SimonSegerblomRex** [comment]
> If I recall how the Pt works, it is basically a right shift of the input samples, so what I'm asking is whether it would be roughly equivalent to compress:
> 
> * 16-bit-wide samples, with only the right-most 14 bits used, as a 14-bit lossless JPEG image and
> * 16-bit-wide samples, with only the left-most 14 bits used, as a 16-bit lossless JPEG image with Pt=2.

I haven't experimented with the point transform, but if I read the specification correctly you could set Pt to 2 if you have 14 bit image data that at some point has been left-shifted 2 bits (so that the 2 lowest bits are always 0) and P to 16 in this case. So you have the two options of:
1. Encoding your "16" bit values and use Pt=2, P=16.
2. Right shifting your values 2 bits prior to encoding and use P=0, P=14.

Both options should produce the same bit streams, so I guess you make the decision based on how you want the values to be decoded (automatically upshifted to 16 bits in case 1, or as 14 bit values in case 2).

**Comment by dcommander** [comment]
That seems to be the case.  If I modify the 16-bit PPM reader so that it shifts rather than scales the input samples to the target data precision:

```
--- a/src/rdppm.c
+++ b/src/rdppm.c
@@ -836,7 +836,10 @@ start_input_ppm(j_compress_ptr cinfo, cjpeg_source_ptr sinfo)
 
   /* Compute the rescaling array if required. */
   if (need_rescale) {
-    long val, half_maxval;
+    long val;
+#if BITS_IN_JSAMPLE != 16
+    long half_maxval;
+#endif
 
     /* On 16-bit-int machines we have to be careful of maxval = 65535 */
     source->rescale = (_JSAMPLE *)
@@ -845,12 +848,19 @@ start_input_ppm(j_compress_ptr cinfo, cjpeg_source_ptr sinfo)
                                            sizeof(_JSAMPLE)));
     memset(source->rescale, 0, (size_t)(((long)MAX(maxval, 255) + 1L) *
                                         sizeof(_JSAMPLE)));
+#if BITS_IN_JSAMPLE != 16
     half_maxval = maxval / 2;
+#endif
     for (val = 0; val <= (long)maxval; val++) {
+#if BITS_IN_JSAMPLE == 16
+      source->rescale[val] =
+        (_JSAMPLE)(val >> (16 - cinfo->data_precision));
+#else
       /* The multiplication here must be done in 32 bits to avoid overflow */
       source->rescale[val] =
         (_JSAMPLE)((val * ((1 << cinfo->data_precision) - 1) + half_maxval) /
                    maxval);
+#endif
     }
   }
 }
```

then the lossless JPEG images generated by
```
cjpeg -lossless 1,2 -precision 16 {input_file}
```
and
```
cjpeg -lossless 1 -precision 14 {input_file}
```
from 16-bit PPM input files differ only in the data precision and Al (point transform) bytes in the JPEG header.

**Comment by dcommander** [comment]
I cannot confirm what you said in https://github.com/libjpeg-turbo/libjpeg-turbo/issues/768#issuecomment-2196487667.  If I lie to the library and tell it that I am compressing samples with 13-bit through 16-bit data precision when I am actually compressing samples with 12-bit data precision (in other words, if I don't rescale the samples in the PPM reader), then compressing lossless JPEG images with 12-bit through 16-bit data precision from the samples with 12-bit data precision produces similar but not identical results.  However, if I do the right thing and scale the input samples to the target data precision, then the lossless JPEG size grows with the data precision.  Am I missing something?  It seems as if the point transform is the only way to make what you said true.

**Comment by SimonSegerblomRex** [comment]
> I cannot confirm what you said in [#768 (comment)](https://github.com/libjpeg-turbo/libjpeg-turbo/issues/768#issuecomment-2196487667). If I lie to the library and tell it that I am compressing samples with 13-bit through 16-bit data precision when I am actually compressing samples with 12-bit data precision (in other words, if I don't rescale the samples in the PPM reader), then compressing lossless JPEG images with 12-bit through 16-bit data precision from the samples with 12-bit data precision produces similar but not identical results. However, if I do the right thing and scale the input samples to the target data precision, then the lossless JPEG size grows with the data precision. Am I missing something? It seems as if the point transform is the only way to make what you said true.

If you have an image with 12 bit values and compress that
1. with `cinfo.data_precision = 12` using `jpeg12_write_scanlines`, or,
2. with `cinfo.data_precision = 13` using `jpeg16_write_scanlines`

you'll recover the exactly the same pixel values when decoding the images. (If you just look at the PGM images the image encoded with the second method will appear darker since the `maxval` value in the header is wrong, but the pixel values are the same.)
